File: packages/python-sdk/src/urisocial/resources/drafts.py
"""Drafts resource"""

from typing import TYPE_CHECKING, List, Optional
from ..types import Draft, PaginatedResponse, PlatformContent

if TYPE_CHECKING:
    from ..http_client import HTTPClient


class DraftsResource:
    """Resource for draft management operations"""

    def __init__(self, http: "HTTPClient"):
        self._http = http

    def list(self, page: int = 1, per_page: int = 20) -> PaginatedResponse:
        """
        Get all drafts with pagination

        Backed by /social-media/content-calendar, brand-scoped under
        multi-tenant mode. The legacy /api/v1/drafts endpoint this used to
        call only ever scoped by the developer's shared account, so every
        end-user saw every other end-user's drafts.

        Args:
            page: Page number (default: 1)
            per_page: Items per page (default: 20)

        Returns:
            Paginated list of drafts
        """
        skip = (page - 1) * per_page
        response = self._http.get(
            "/social-media/content-calendar",
            params={"skip": skip, "limit": per_page},
        )
        data = response["responseData"]
        return {
            "data": data["drafts"],
            "total": data["total_count"],
            "page": page,
            "per_page": per_page,
            "has_more": data["pagination"]["has_more"],
        }

    def get(self, draft_id: str) -> Draft:
        """
        Get a specific draft by ID

        Args:
            draft_id: ID of the draft

        Returns:
            Draft details
        """
        response = self._http.get(f"/social-media/drafts/{draft_id}")
        return response["responseData"]

    def update(
        self,
        draft_id: str,
        text_content: Optional[List[PlatformContent]] = None,
        image_url: Optional[str] = None,
    ) -> Draft:
        """
        Update draft content

        Args:
            draft_id: ID of the draft to update
            text_content: Updated platform-specific text content
            image_url: Updated image URL

        Returns:
            Updated draft
        """
        updates = {}
        if text_content is not None:
            updates["text_content"] = text_content
        if image_url is not None:
            updates["image_url"] = image_url

        response = self._http.patch(f"/social-media/drafts/{draft_id}", json=updates)
        return response["responseData"]

    def delete(self, draft_id: str) -> dict:
        """
        Delete a draft

        Args:
            draft_id: ID of the draft to delete

        Returns:
            Success confirmation
        """
        return self._http.delete(f"/social-media/drafts/{draft_id}")

    def create(
        self,
        text_content: List[PlatformContent],
        image_url: Optional[str] = None,
        reference_image: Optional[str] = None,
    ) -> Draft:
        """
        Create a new draft manually

        Not currently supported — no brand-scoped equivalent of the legacy
        /api/v1/drafts create endpoint exists yet. That legacy endpoint is
        not brand-isolated under multi-tenant mode, so it's intentionally
        not called here. Use content.generate() to create drafts instead.
        """
        raise NotImplementedError(
            "drafts.create() is not currently supported for SDK/multi-tenant use — "
            "use content.generate() instead."
        )

File: packages/typescript-sdk/src/resources/drafts.ts
import { HTTPClient } from '../client/http';
import { Draft, PaginatedResponse, PlatformContent } from '../types';

export interface RegenerateImageRequest {
  prompt?: string;
  style?: string;
}

export interface EditImageRequest {
  prompt: string;
  mask_url?: string;
}

export interface UpdateSlideRequest {
  headline?: string;
  body?: string;
  image_url?: string;
}

export class DraftsResource {
  constructor(private http: HTTPClient) {}

  /**
   * Get all drafts with pagination
   *
   * @remarks Backed by /social-media/content-calendar, brand-scoped under
   * multi-tenant mode — the legacy /api/v1/drafts endpoint this used to call
   * only ever scoped by the developer's shared account, so every end-user
   * saw every other end-user's drafts.
   */
  async list(page: number = 1, perPage: number = 20): Promise<PaginatedResponse<Draft>> {
    const skip = (page - 1) * perPage;
    const response = await this.http.get<{
      responseData: {
        drafts: Draft[];
        total_count: number;
        pagination: { skip: number; limit: number; has_more: boolean };
      };
    }>(`/social-media/content-calendar?skip=${skip}&limit=${perPage}`);

    const { drafts, total_count, pagination } = response.responseData;
    return {
      data: drafts,
      total: total_count,
      page,
      per_page: perPage,
      has_more: pagination.has_more,
    } as PaginatedResponse<Draft>;
  }

  /**
   * Get a specific draft by ID
   */
  async get(draftId: string): Promise<Draft> {
    return this.http.get<Draft>(`/social-media/drafts/${draftId}`);
  }

  /**
   * Get draft image URL (for streaming/progressive loading)
   */
  async getImage(draftId: string): Promise<{ image_url?: string; status: string }> {
    return this.http.get<{ image_url?: string; status: string }>(
      `/social-media/draft-image/${draftId}`
    );
  }

  /**
   * Update draft content
   */
  async update(
    draftId: string,
    updates: {
      text_content?: PlatformContent[];
      image_url?: string;
    }
  ): Promise<Draft> {
    const response = await this.http.patch<{ responseData: Draft }>(
      `/social-media/drafts/${draftId}`,
      updates
    );
    return response.responseData;
  }

  /**
   * Delete a draft
   */
  async delete(draftId: string): Promise<{ success: boolean }> {
    return this.http.delete<{ success: boolean }>(`/social-media/drafts/${draftId}`);
  }

  /**
   * Create a new draft manually
   *
   * @remarks Not currently supported — no brand-scoped equivalent of the
   * legacy /api/v1/drafts create endpoint exists yet. That legacy endpoint
   * is not brand-isolated under multi-tenant mode, so it's intentionally
   * not called here. Use content.generate() to create drafts instead.
   */
  async create(_draft: {
    text_content: PlatformContent[];
    image_url?: string;
    reference_image?: string;
  }): Promise<Draft> {
    throw new Error(
      'drafts.create() is not currently supported for SDK/multi-tenant use — use content.generate() instead.'
    );
  }

  /**
   * Regenerate image for a draft
   * @param draftId - Draft ID
   * @param request - Image regeneration parameters
   */
  async regenerateImage(
    draftId: string,
    request?: RegenerateImageRequest
  ): Promise<{ image_url: string }> {
    return this.http.post<{ image_url: string }>(
      `/social-media/drafts/${draftId}/regenerate-image`,
      request || {}
    );
  }

  /**
   * Edit draft image with AI
   * @param draftId - Draft ID
   * @param request - Image edit parameters
   */
  async editImage(draftId: string, request: EditImageRequest): Promise<{ image_url: string }> {
    return this.http.post<{ image_url: string }>(
      `/social-media/drafts/${draftId}/edit-image`,
      request
    );
  }

  /**
   * Undo last image edit
   * @param draftId - Draft ID
   */
  async undoImageEdit(draftId: string): Promise<{ image_url: string }> {
    return this.http.post<{ image_url: string }>(
      `/social-media/drafts/${draftId}/undo-image`
    );
  }

  /**
   * Unschedule a scheduled draft
   * @param draftId - Draft ID
   */
  async unschedule(draftId: string): Promise<{ success: boolean }> {
    return this.http.post<{ success: boolean }>(
      `/social-media/drafts/${draftId}/unschedule`
    );
  }

  // ============================================================================
  // Carousel-specific methods
  // ============================================================================

  /**
   * Update a specific slide in a carousel draft
   * @param draftId - Draft ID
   * @param slideIndex - Slide index (0-based)
   * @param updates - Slide updates
   */
  async updateSlide(
    draftId: string,
    slideIndex: number,
    updates: UpdateSlideRequest
  ): Promise<{ slide: any }> {
    return this.http.patch<{ slide: any }>(
      `/social-media/drafts/${draftId}/slides/${slideIndex}`,
      updates
    );
  }

  /**
   * Regenerate image for a specific slide
   * @param draftId - Draft ID
   * @param slideIndex - Slide index
   */
  async regenerateSlideImage(draftId: string, slideIndex: number): Promise<{ image_url: string }> {
    return this.http.post<{ image_url: string }>(
      `/social-media/drafts/${draftId}/slides/${slideIndex}/regenerate-image`
    );
  }

  /**
   * Delete a slide from carousel
   * @param draftId - Draft ID
   * @param slideIndex - Slide index
   */
  async deleteSlide(draftId: string, slideIndex: number): Promise<{ success: boolean }> {
    return this.http.delete<{ success: boolean }>(
      `/social-media/drafts/${draftId}/slides/${slideIndex}`
    );
  }

  /**
   * Reorder slides in a carousel
   * @param draftId - Draft ID
   * @param newOrder - Array of slide indices in new order
   */
  async reorderSlides(draftId: string, newOrder: number[]): Promise<{ success: boolean }> {
    return this.http.post<{ success: boolean }>(
      `/social-media/drafts/${draftId}/slides/reorder`,
      { new_order: newOrder }
    );
  }
}
